### PR TITLE
Parsing Error Detection

### LINKED ISSUE #26: Add error detection in parser and parsing table.
# Feature Request
## Summary
Add error detection in Parser and ParsingTable

## Solution
We need to add more error statements in Parser class.
Add support in ParsingTable to detect grammar related errors like non ll(1) grammar.

Checklist
- [ ] Add error statements in Parser class
- [ ] Add error detection in ParsingTable class
- [ ] Make sure incorrect grammar file or input file doesn't cause segmentation faults.

### PR DESCRIPTION
**Pull request by TheSYNcoder**
# Parsing and Parsing Table Error Detection.
Fixes Error Detection problem in construction of ll1 parsing table and the parser.

Errors are now detected in case of non-LL(1) grammar and the input is not parsed until the grammar provided is LL(1).

Checklist

- [X] Add error statements in Parser class
- [X] Add error detection in ParsingTable class
- [X] Make sure incorrect grammar file or input file doesn't cause segmentation faults.

Fixes #26

### CODE REVIEW

File: src/include/parser/parser.h [hunk 1]
```diff
@@ -49,6 +50,17 @@ class Parser {
    */
   std::vector<std::string> current_string_;
 
+  /**
+   * Errors incurred, if any, during the construction of the
+   * LL(1) parsing table.
+   */
+  std::vector<std::string> parsing_table_errors_;
```
Comment (noob77777) [inline src/include/parser/parser.h]: why store parsing_table_errors_ inside parser? This should be handled in main.

File: src/main/main.cpp [hunk 2]
```diff
@@ -20,6 +20,8 @@
  *-------------------------------------------------------------------------
  */
 
+#include <iostream>
```
Comment (noob77777) [inline src/main/main.cpp]: not required

File: src/main/main.cpp [hunk 3]
```diff
@@ -81,9 +83,22 @@ auto main(int argc, char *argv[]) -> int {
   parser.SetStartSymbol(grammar_parser.GetStartSymbol());
   parser.SetParsingTable(parsing_table);
 
+  auto err = parsing_table.GetErrors();
+  if (!err.empty()) {
+    for (auto &e : err) {
+      std::cout << e << "\n";
+      return 0;
```
Comment (noob77777) [inline src/main/main.cpp]: If there are errors we are already returning 0. So why does Parser class store a copy of parsing_table_errors?

File: src/parser/parser.cpp [hunk 4]
```diff
@@ -50,11 +50,22 @@ void Parser::DoNextStep() {
 }
 
 void Parser::ParseNextStep() {
+  // if there are parsing table errors do not perform parsing
+  if (!parsing_table_errors_.empty()) {
```
Comment (noob77777) [inline src/parser/parser.cpp]: this check is not required.

File: src/parser/parser.cpp [hunk 5]
```diff
@@ -50,11 +50,22 @@ void Parser::DoNextStep() {
...
+    parser_errors_ = parsing_table_errors_;
+    return;
+  }
   std::string top_symbol = stack_.top();
   std::string current_token = current_string_[current_step_];
   ParsingTable::Table table = table_.GetTable();
   // skip tokens until it is in the first or is a synch token
   while (!IsComplete() && table[top_symbol][current_token] == std::string(ERROR_TOKEN)) {
+    std::string ret;
+    ret += "Parsing error at symbol : " + current_token + " \n";
```
Comment (noob77777) [inline src/parser/parser.cpp]: According to convention should be "parser: error at symbol: " + current_token + "\n";

File: src/parser/parser.cpp [hunk 6]
```diff
@@ -63,6 +74,9 @@ void Parser::ParseNextStep() {
   if (!IsComplete()) {
     // if SYNCH TOKEN - We skip the current symbol on stack top
     if (table[top_symbol][current_token] == std::string(SYNCH_TOKEN)) {
+      std::string ret;
+      ret += "Parsing error at symbol : " + current_token + " \n";
```
Comment (noob77777) [inline src/parser/parser.cpp]: Same

File: src/parser/parsing_table.cpp [hunk 7]
```diff
@@ -17,6 +17,13 @@ void ParsingTable::BuildTable() {
   for (auto &nt : non_terminals_) {
     if (follows_.count(nt) != 0U) {
       for (const auto &symbol : follows_[nt]) {
+        if (table_[nt][symbol] != std::string(ERROR_TOKEN)) {
+          std::string ret;
+          ret += "Error duplicate entry in parsing table, ";
```
Comment (noob77777) [inline src/parser/parsing_table.cpp]: "error: duplicate entry ..."
First character is not capitalized in unix systems generally.
no space before ":"
should be "production: " and not "Production : " etc.

File: src/parser/parsing_table.cpp [hunk 8]
```diff
@@ -30,29 +37,47 @@ void ParsingTable::BuildTable() {
 
       // check if first_entity is terminal
       if (std::find(terminals_.begin(), terminals_.end(), first_entity) != terminals_.end()) {
+        std::string entry = table_[productions_[prod_no].GetParent()][first_entity];
+        if ((entry != std::string(ERROR_TOKEN)) && (entry != std::string(SYNCH_TOKEN))) {
+          std::string ret;
+          ret += "Error duplicate entry in parsing table, ";
+          ret += "Production : " + productions_[prod_no].GetParent();
+          ret += " Symbol : " + first_entity;
```
Comment (noob77777) [inline src/parser/parsing_table.cpp]: same

File: src/parser/parsing_table.cpp [hunk 9]
```diff
@@ -30,29 +37,47 @@ void ParsingTable::BuildTable() {
...
-          // if epsilon present add production under follow(A) symbols
-          if (std::find(firsts_[first_entity].begin(), firsts_[first_entity].end(), std::string(grammar::EPSILON)) !=
-              firsts_[first_entity].end()) {
-            if (follows_.count(productions_[prod_no].GetParent()) != 0U) {
-              for (auto &symbol : follows_[productions_[prod_no].GetParent()]) {
-                table_[productions_[prod_no].GetParent()][symbol] = std::to_string(prod_no * 100 + rule_no);
+              std::string entry = table_[productions_[prod_no].GetParent()][symbol];
+              if ((entry != std::string(ERROR_TOKEN)) && (entry != std::string(SYNCH_TOKEN))) {
+                std::string ret;
+                ret += "Error duplicate entry in parsing table, ";
```
Comment (noob77777) [inline src/parser/parsing_table.cpp]: same

File: src/parser/parsing_table.cpp [hunk 10]
```diff
@@ -30,29 +37,47 @@ void ParsingTable::BuildTable() {
...
         }
+
       } else if (first_entity == std::string(grammar::EPSILON)) {
         // first entity is epsilon
         if (follows_.count(productions_[prod_no].GetParent()) != 0U) {
           for (auto &symbol : follows_[productions_[prod_no].GetParent()]) {
+            std::string entry = table_[productions_[prod_no].GetParent()][symbol];
+            if ((entry != std::string(ERROR_TOKEN)) && (entry != std::string(SYNCH_TOKEN))) {
+              std::string ret;
+              ret += "Error duplicate entry in parsing table, ";
```
Comment (noob77777) [inline src/parser/parsing_table.cpp]: Create a separate function to generate this message.
Duplicate code used four times.

File: test/parser/parsing_table_test.cpp [hunk 11]
```diff
@@ -1,5 +1,7 @@
 #include "parser/parsing_table.h"
 
+#include <iostream>
```
Comment (noob77777) [inline test/parser/parsing_table_test.cpp]: not required

File: test/parser/parsing_table_test.cpp [hunk 12]
```diff
@@ -250,3 +264,105 @@ TEST(parser, ParsingTable4) {
...
+  std::vector<std::string> terminals = {"a", "b"};
+  std::vector<std::string> non_terminals = {"A", "S", "B"};
+
+  ParsingTable table = ParsingTable(terminals, non_terminals);
+  table.SetFirsts(firsts);
+  table.SetFollows(follows);
+  table.SetProductions(grammar);
+  table.BuildTable();
+
+  ASSERT_GT(table.GetErrors().size(), 0);
```
Comment (noob77777) [inline test/parser/parsing_table_test.cpp]: ASSERT_EQ
test for exact number of errors

File: test/parser/parsing_table_test.cpp [hunk 13]
```diff
@@ -250,3 +264,105 @@ TEST(parser, ParsingTable4) {
...
+  std::vector<std::string> terminals = {"a"};
+  std::vector<std::string> non_terminals = {"A", "S"};
+
+  ParsingTable table = ParsingTable(terminals, non_terminals);
+  table.SetFirsts(firsts);
+  table.SetFollows(follows);
+  table.SetProductions(grammar);
+  table.BuildTable();
+
+  ASSERT_GT(table.GetErrors().size(), 0);
```
Comment (noob77777) [inline test/parser/parsing_table_test.cpp]: ASSERT_EQ

File: test/parser/parsing_table_test.cpp [hunk 14]
```diff
@@ -40,6 +42,9 @@ TEST(parser, ParsingTable1) {
   table.SetProductions(grammar);
   table.BuildTable();
 
+  std::vector<std::string> tl;
+  ASSERT_EQ(table.GetErrors(), tl);
```
Comment (noob77777) [inline test/parser/parsing_table_test.cpp]: Compare size to 0

**Review by noob77777 — CHANGES_REQUESTED** [review]

File: src/parser/parser.cpp (hunk 4 above)
Comment (TheSYNcoder) [reply src/parser/parser.cpp]: Why parsing would not be performed if there is any errors in parsing_table
The input grammar should be correct!

File: src/parser/parser.cpp (hunk 5 above)
Comment (TheSYNcoder) [reply src/parser/parser.cpp]: Yeah okay!

File: src/parser/parser.cpp (hunk 4 above)
Comment (noob77777) [reply src/parser/parser.cpp]: because the compiler already returns 0 in main. Having parsing_table_errors_ inside Parser class in bad design.

Comment (TheSYNcoder) [reply src/parser/parser.cpp]: Ahh Okay.

Comment (noob77777) [reply src/parser/parser.cpp]: Else do parsing_table_.GetErrors() or something having a extra variable is completely unnecessary.

File: src/main/main.cpp [hunk 15]
```diff
@@ -81,9 +83,22 @@ auto main(int argc, char *argv[]) -> int {
...
+    }
+  }
   while (!parser.IsComplete()) {
     parser.ParseNextStep();
   }
+  err = parser.GetParserErrors();
+  if (!err.empty()) {
+    for (auto &e : err) {
```
Comment (bisakhmondal) [inline src/main/main.cpp]: Redundant checking. Not required

File: src/main/main.cpp [hunk 16]
```diff
@@ -81,9 +83,22 @@ auto main(int argc, char *argv[]) -> int {
   parser.SetStartSymbol(grammar_parser.GetStartSymbol());
   parser.SetParsingTable(parsing_table);
 
+  auto err = parsing_table.GetErrors();
+  if (!err.empty()) {
+    for (auto &e : err) {
```
Comment (bisakhmondal) [inline src/main/main.cpp]: same

File: src/parser/parsing_table.cpp (hunk 10 above)
Comment (bisakhmondal) [reply src/parser/parsing_table.cpp]: yes agree +1

**Review by bisakhmondal — COMMENTED** [review]
Rest is good

File: test/parser/parsing_table_test.cpp [hunk 17]
```diff
@@ -158,6 +166,9 @@ TEST(parser, ParsingTable3) {
...
+
```
Comment (bisakhmondal) [inline test/parser/parsing_table_test.cpp]: for a single instance new variable is not required.
use `std::vector<std::string>{}`

**Comment by TheSYNcoder** [comment]
If you want further changes write yourself,
I want to see this merged by tomorrow.
Going to sleep.
Good Night.

**Review by noob77777 — APPROVED** [review]

**Pull request merged**